File: matlab/smop.m
function res = smop(A, b, delta, opts)
%BMOP  Solve min ||x||_1  s.t.  ||A x - b||_2 <= delta (level-set method).
%
%   res = BMOP(A, b, delta)
%   res = BMOP(A, b, delta, opts)
%
%   A      : m x n dense double matrix
%   b      : m x 1 vector
%   delta  : scalar constraint radius
%   opts   : (optional) struct with fields
%              stoptol   (1e-6)  outer tolerance
%              maxiter   (200)   max level-set iterations
%              use_secant (true) secant root-finding steps
%              use_newton (false) analytic-Newton steps
%              use_as     (true) adaptive sieving for subproblems
%              mu0, muinf, initial_mu : bracket / initial point
%              verbose   (0)
%              time_limit (3600)
%
%   res is a struct with fields
%     x, xi, y, mu, psi, eta, iter, iter_bisection,
%     iter_newton_or_secant, iter_smoothing, iter_admm, status, msg,
%     mupath, psipath, reducedn
%
% Example:
%   rng(1); A = randn(40,150); xt = zeros(150,1);
%   xt(randperm(150,8)) = randn(8,1); b = A*xt + 0.02*randn(40,1);
%   delta = 1.02*norm(A*xt - b);
%   res = smop(A, b, delta); fprintf('||Ax-b|| = %g, delta = %g\n', norm(A*res.x-b), delta);
%
% See also: smop_lasso, build_smop
%
% Build the MEX file first (once) with: build_smop

if nargin < 3, error('smop:arg', 'usage: smop(A, b, delta [, opts])'); end
if nargin < 4, opts = struct(); end
res = smop_mex(A, b, delta, opts);
end

File: R/R/smop.R
#======================================================================
# smop.R -- R-level wrappers for the smop C++ core (Rcpp bindings)
#======================================================================

#' Solve the BMOP (Basis-pursuit with a level-set constraint):
#'   min ||x||_1  s.t.  ||A x - b||_2 <= delta
#' by the level-set method (secant / Newton / bisection + adaptive sieving).
#'
#' @param A      an m x n numeric matrix (dense)
#' @param b      numeric vector of length m
#' @param delta  positive scalar constraint radius
#' @param options optional named list, e.g.
#'   list(stoptol=1e-6, maxiter=200, use_secant=TRUE, use_as=TRUE, verbose=0)
#' @param x0     optional warm start vector
#' @return a list with elements x, xi, y, mu, psi, eta, iter, status, msg,
#'   time, mupath, psipath, reducedn
#' @export
solve_bmop <- function(A, b, delta, options = NULL, x0 = NULL) {
  if (!is.matrix(A)) A <- as.matrix(A)
  b <- as.numeric(b)
  stopifnot(ncol(A) == length(b) || nrow(A) == length(b))
  if (nrow(A) != length(b)) A <- t(A)          # column convention
  smop_solve_bmop(A, b, delta, options, x0)
}

#' Solve the Lasso subproblem:
#'   min 1/2 ||A x - b||_2^2 + lambda ||x||_1
#' @export
solve_lasso <- function(A, b, lambda, options = NULL, x0 = NULL) {
  if (!is.matrix(A)) A <- as.matrix(A)
  b <- as.numeric(b)
  stopifnot(ncol(A) == length(b) || nrow(A) == length(b))
  if (nrow(A) != length(b)) A <- t(A)
  smop_solve_lasso(A, b, lambda, options, x0)
}

#' Package version.
#' @export
smop_version <- function() smop_version_cpp()
